lib/renderer/init.ts

```diff
@@ -122,15 +122,27 @@ if (nodeIntegration) {
   }
 }
 
+const { appCodeLoaded } = process;
+delete process.appCodeLoaded;
+
 const { preloadPaths } = ipcRendererUtils.invokeSync(IPC_MESSAGES.BROWSER_NONSANDBOX_LOAD);
+const preloadPromises: Promise<void>[] = [];
 // Load the preload scripts.
 for (const preloadScript of preloadPaths) {
-  try {
-    Module._load(preloadScript);
-  } catch (error) {
-    console.error(`Unable to load preload script: ${preloadScript}`);
-    console.error(error);
+  // Finally load app's main.js and transfer control to C++.
+  const { loadESM } = __non_webpack_require__('internal/process/esm_loader');
+  preloadPromises.push(loadESM((esmLoader: any) => {
+    return esmLoader.import(preloadScript, undefined, Object.create(null)).catch((err: Error) => {
+      console.error(`Unable to load preload script: ${preloadScript}`);
+      console.error(err);
+
+      ipcRendererInternal.send(IPC_MESSAGES.BROWSER_PRELOAD_ERROR, preloadScript, err);
+    });
+  }));
+}
 
-    ipcRendererInternal.send(IPC_MESSAGES.BROWSER_PRELOAD_ERROR, preloadScript, error);
-  }
+if (preloadPromises.length) {
+  Promise.all(preloadPromises).then(() => appCodeLoaded!());
+} else {
+  appCodeLoaded!();
 }
```

Comment:
Kind of silly of that `loadESM` is an async function that takes a callback...

I think that this segment could be meaningfully simplified:

* `loadESM` doesn't need to be imported & called every iteration. Both can be hoisted out of the loop _afaict_.
* `Promise.all` still works fine with an empty array. Are you trying to avoid queuing a microtask in the case where there are no preload scripts?

Reply:
> Are you trying to avoid queuing a microtask in the case where there are no preload scripts?

yes

> Both can be hoisted out of the loop afaict.

I'll have to double check but I think you get a new esm loader each time you call this 🤔 Could be wrong though, will check tomorrow